lib/kafka/async_producer.rb

```diff
@@ -0,0 +1,117 @@
+require "thread"
+
+module Kafka
+
+  # A Kafka producer that does all its work in the background so as to not block
+  # the calling thread. Calls to {#deliver_messages} are asynchronous and return
+  # immediately.
+  #
+  # In addition to this property it's possible to define automatic delivery
+  # policies. These allow placing an upper bound on the number of buffered
+  # messages and the time between message deliveries.
+  #
+  # * If `delivery_threshold` is set to a value _n_ higher than zero, the producer
+  #   will automatically deliver its messages once its buffer size reaches _n_.
+  # * If `delivery_interval` is set to a value _n_ higher than zero, the producer
+  #   will automatically deliver its messages every _n_ seconds.
+  #
+  # ## Example
+  #
+  #     producer = kafka.get_async_producer(
+  #       # Keep at most 1.000 messages in the buffer before delivering:
+  #       delivery_threshold: 1000,
+  #
+  #       # Deliver messages every 30 seconds:
+  #       delivery_interval: 30,
+  #     )
+  #
+  #     # There's no need to manually call #deliver_messages, it will happen
+  #     # automatically in the background.
+  #     producer.produce("hello", topic: "greetings")
+  #
+  class AsyncProducer
+    def initialize(sync_producer:, delivery_threshold: 0, delivery_interval:)
```

Comment:
I'd like to see defaults here that aren't 0/nil. Maybe just 16 buffered messages and a delivery interval of 0? The Go producer has a 0 here, and it means every single production use we've done with it runs into issues until they tweak it.

The JVM and Go Kafka producers use bytes instead of number of messages here, and I think that's strictly superior - the reason you want to limit the max number of pending things is to prevent your memory from bloating to infinity when a runaway process produces too many things. (Your task about limiting the queue also has impacts there).

From a design perspective, the JVM producer also lets you pass a callback to track errors and successes. I guess in `ruby-kafka` the intention there is to just use the inbuilt metrics mechanism? I think that's notably worse than being able to pass a callback - a callback lets you do specific things on each success/failure, and in my experience that's often very useful (e.g. you may want to block the response of an http request until the kafka message it produced has been sent).

I don't think all of the above has to land in this PR, but I think they're important to making the producer operationally stable and usable without tuning.